### PR TITLE
Add `CoordsIter` trait for iterating over coordinates in geometries.

### PR DESCRIPTION
**Pull request by frewsxcv**
A new trait to iterate over all coordinates in a geometry. In the future, we should add a new method to this trait called `external_coords_iter` which iterates over the external coordinates of a geometry, excluding coordinates of interior holes. Then we can make algorithms like `ConvexHull` have a generic implementation where `G: CoordsIter`

### CODE REVIEW

**Comment by frewsxcv** [comment]
working on refactoring some of the rust-geo traits right now. this is similar to `MapCoords` but unlike `MapCoords` the methods on this trait don't result in any new memory allocations.

**Comment by rory** [comment]
I have a `map_coords_inplace` branch lying around at home that I should make a PR for. It allows you to change things without memory allocations But maybe this would suffice.

**Comment by frewsxcv** [comment]
> It allows you to change things without memory allocations But maybe this would suffice.

yep, the main reason i opened this is so that i can modify `Coordinate` values without allocations:

https://github.com/georust/rust-geo/compare/frewsxcv-rotate-refactgor?expand=1#diff-64481126a8d2c515ad92dc33cc71a856R18

**Comment by rory** [comment]
:) I also wanted to modify objects without allocations.

I should clean up my code and start opening PRs

**Comment by frewsxcv** [comment]
@rory don't want to pressure you into reviewing, but did you happen to already look at the code here? if so, any thoughts?

**Comment by rory** [comment]
I only skimmed it earlier. I might be able to look at it properly within the next few days

File: src/algorithm/coords_iter.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+use num_traits::Float;
+use std::iter;
+use types::{Coordinate, Point, Polygon, LineString, Line, MultiPoint, MultiPolygon,
+            MultiLineString};
+
+pub trait CoordsIter<T: Float> {
+    fn coords_iter<'a>(&'a self) -> Box<Iterator<Item = &'a Coordinate<T>> + 'a>;
```
Comment (frewsxcv) [inline src/algorithm/coords_iter.rs]: just like the other boxed iterators that get returned in this crate, we can remove the `Box` once return `impl`s stabilize

**Comment by frewsxcv** [comment]
just finished this! i'll try to find some time in the next few days to add tests. most of the time i spent on this was trying to remove `Box<Iterator>`, which is why we have to specify the specific types of the iterator for each implementation

**Comment by frewsxcv** [comment]
also will add docs to the trait. just realized they're missing

**Review by michaelkirk — APPROVED** [review]
nice!

**Comment by frewsxcv** [comment]
bors r=michaelkirk

**Pull request merged**